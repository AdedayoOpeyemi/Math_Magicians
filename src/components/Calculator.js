import React from 'react';
import calculate from '../logic/calculate';

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      total: null,
      next: null,
      operation: null,
    };
    this.clickHandler = this.clickHandler.bind(this);
    this.displayResult = this.displayResult.bind(this);
  }

  clickHandler(e) {
    this.setState((objName) => calculate(objName, e.target.dataset.buttonName));
  }

  displayResult() {
    const { total, next, operation } = this.state;
    let result = null;

    if (total === null && next === null) {
      result = 0;
    } else if (next !== null && total === null) {
      result = next;
    } else if (next !== null && total !== null) {
      if (operation !== null && operation !== undefined) {
        result = `${total} ${operation} ${next}`;
      } else {
        result = total;
      }
    } else if (next === null && total !== null) {
      if (operation !== null && operation !== undefined) {
        result = `${total} ${operation}`;
      } else {
        result = total;
      }
    }
    return result;
  }

  render() {
    return (
      <div className="calc-container">
        <div className="calc_display">{ this.displayResult() }</div>
        <div className="int_values" role="button" data-button-name="AC" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>AC</div>
        <div className="int_values" role="button" data-button-name="+/-" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>+/-</div>
        <div className="int_values" role="button" data-button-name="%" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>%</div>
        <div className="arithmetic_operand" role="button" data-button-name="รท" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>รท</div>
        <div className="int_values" role="button" data-button-name="1" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>1</div>
        <div className="int_values " role="button" data-button-name="2" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>2</div>
        <div className="int_values " role="button" data-button-name="3" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>3</div>
        <div className="arithmetic_operand" role="button" data-button-name="x" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>x</div>
        <div className="int_values " role="button" data-button-name="4" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>4</div>
        <div className="int_values " role="button" data-button-name="5" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>5</div>
        <div className="int_values " role="button" data-button-name="6" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>6</div>
        <div className="arithmetic_operand" role="button" data-button-name="-" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>-</div>
        <div className="int_values " role="button" data-button-name="7" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>7</div>
        <div className="int_values " role="button" data-button-name="8" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>8</div>
        <div className="int_values " role="button" data-button-name="9" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>9</div>
        <div className="arithmetic_operand" role="button" data-button-name="+" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>+</div>
        <div className="int_values zero" role="button" data-button-name="0" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>0</div>
        <div className="int_values " role="button" data-button-name="0" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>.</div>
        <div className="arithmetic_operand" role="button" data-button-name="=" onClick={this.clickHandler} onKeyDown={this.clickHandler} tabIndex={0}>=</div>
      </div>
    );
  }
}

export default Calculator;
