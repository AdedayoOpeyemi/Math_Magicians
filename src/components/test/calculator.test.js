import { render, screen, fireEvent, getByLabelText } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Calculator from '../Calculator';

it('renders correctly', () => {
  const tree = renderer
    .create(<Calculator />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders the changes when the calcultor button is clicked', () => {
  render(<Calculator />);
  fireEvent.click(screen.getByText('1'));
  fireEvent.click(screen.getByText('5'));
  const typedElement = screen.getByText('15');
  expect(typedElement.innerHTML).toBe('15');
});

test('renders the changes when the calcultor button is clicked', () => {
  render(<Calculator />);
  fireEvent.click(screen.getByText('1'));
  fireEvent.click(screen.getByText('5'));
  fireEvent.click(screen.getByText('+'));
  fireEvent.click(screen.getByText('3'));
  const typedElement = screen.getByText('15 + 3');
  expect(typedElement.innerHTML).toBe('15 + 3');
});

test('renders the changes when the calcultor button is clicked', () => {
  render(<Calculator />);
  fireEvent.click(screen.getByText('1'));
  fireEvent.click(screen.getByText('5'));
  fireEvent.click(screen.getByText('+'));
  fireEvent.click(screen.getByText('3'));
  let typedElement = screen.getByText('15 + 3');
  expect(typedElement.innerHTML).toBe('15 + 3');
  fireEvent.click(screen.getByText('='));
  typedElement = screen.getByText('18');
  expect(typedElement.innerHTML).toBe('18');
});

