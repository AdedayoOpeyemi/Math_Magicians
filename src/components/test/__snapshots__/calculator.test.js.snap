// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="calc-container"
>
  <div
    className="calc_display"
  >
    0
  </div>
  <div
    className="int_values"
    data-button-name="AC"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    AC
  </div>
  <div
    className="other_operators"
    data-button-name="+/-"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    +/-
  </div>
  <div
    className="other_operators"
    data-button-name="%"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    %
  </div>
  <div
    className="arithmetic_operand"
    data-button-name="รท"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    รท
  </div>
  <div
    className="int_values"
    data-button-name="1"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    1
  </div>
  <div
    className="int_values "
    data-button-name="2"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    2
  </div>
  <div
    className="int_values "
    data-button-name="3"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    3
  </div>
  <div
    className="arithmetic_operand"
    data-button-name="x"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    x
  </div>
  <div
    className="int_values "
    data-button-name="4"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    4
  </div>
  <div
    className="int_values "
    data-button-name="5"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    5
  </div>
  <div
    className="int_values "
    data-button-name="6"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    6
  </div>
  <div
    className="arithmetic_operand"
    data-button-name="-"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    -
  </div>
  <div
    className="int_values "
    data-button-name="7"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    7
  </div>
  <div
    className="int_values "
    data-button-name="8"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    8
  </div>
  <div
    className="int_values "
    data-button-name="9"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    9
  </div>
  <div
    className="arithmetic_operand"
    data-button-name="+"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    +
  </div>
  <div
    className="int_values zero"
    data-button-name="0"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    0
  </div>
  <div
    className="int_values "
    data-button-name="."
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    .
  </div>
  <div
    className="arithmetic_operand"
    data-button-name="="
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    =
  </div>
</div>
`;
